cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java

```diff
@@ -138,8 +144,16 @@ public final void reset() {
 
     public Macro(String name, @Nullable List<Token> params, @Nullable List<Token> body, boolean variadic) {
```

Comment:
@Bertk think initial problem is that params and body are marked @Nullable. Think these two parameters should always be valid. Looking to the places with `new Macro(...)` there seems to be only `registerMacros` where params=null. Maybe better to forward there an empty list and adapt Macro class according. In most other cases seems to be a performance penalty where list is duplicated without need.

Reply:
@guwirth unfortunatly there is a nasty nested ternary operator and a logic which uses
```macroParams != null``` -> this method has to be refactored ...

```
   private Macro parseMacroDefinition(AstNode defineLineAst) {
    AstNode ast = defineLineAst.getFirstChild();
    AstNode nameNode = ast.getFirstDescendant(CppGrammar.ppToken);
    String macroName = nameNode.getTokenValue();

    AstNode paramList = ast.getFirstDescendant(CppGrammar.parameterList);
    List<Token> macroParams = paramList == null
      ? "objectlikeMacroDefinition".equals(ast.getName()) ? null : new LinkedList<>() : getParams(paramList);

    AstNode vaargs = ast.getFirstDescendant(CppGrammar.variadicparameter);
    if ((vaargs != null) && (macroParams != null)) {
      AstNode identifier = vaargs.getFirstChild(IDENTIFIER);
      macroParams.add(identifier == null
          ? Token.builder()
          .setLine(vaargs.getToken().getLine())
          .setColumn(vaargs.getToken().getColumn())
          .setURI(vaargs.getToken().getURI())
          .setValueAndOriginalValue(VARIADICPARAMETER)
          .setType(IDENTIFIER)
          .setGeneratedCode(true)
          .build()
          : identifier.getToken());
    }

    AstNode replList = ast.getFirstDescendant(CppGrammar.replacementList);
    List<Token> macroBody = replList == null
      ? new LinkedList<>() : replList.getTokens().subList(0, replList.getTokens().size() - 1);

    return new Macro(macroName, macroParams, macroBody, vaargs != null);
  }
```

Reply:
I will revert the ctor update. This shall be refactored by a separate PR.